### DISCUSSION TITLE
Treating Cyclic Boundary Conditions

### DISCUSSION THREAD

**Discussion by JanGaertner** [WENO-OF/WENOEXT#20] [Ideas]
In the current implementation, the stencil collection algorithm cannot handle cyclic boundary conditions. This means that for cells at at cyclic boundary condition, cells on the other side of the boundary will not be considered.

## Stencil Collection with Cyclic Boundaries

To collect the cells for the stencil the `cellCells()` function is used recursively to get all neighboring cells in `WENOBase::createStencilID()`. This function however, does not return cells beyond any boundary condition even if they are at a cyclic or processor boundary. 
A possibility to collect all cells even at a cyclic boundary condition could be the usage of a getNeighbourCells() function:

```c++
Foam::labelList Foam::WENOBase::getNeighbourCells
(
    const fvMesh& mesh,
    const label cellI
)
{
    labelList neighbCells = mesh.cellCells()[cellI];
    
    // Check if cell is at the boundary. 
    const cell& faces = mesh.cells()[cellI];
    const polyBoundaryMesh& bMesh = mesh.boundaryMesh();
    forAll(faces,faceI)
    {
        // Internal faces return -1 other faces return boundaryPatchID
        const label boundaryPatchID = bMesh.whichPatch(faces[faceI]);
        if (boundaryPatchID != -1)
        {
            if (   bMesh[boundaryPatchID].coupled() 
                && isA<cyclicPolyPatch>(bMesh[boundaryPatchID]) )
            {
                const cyclicPolyPatch& patch = refCast<const cyclicPolyPatch>
                    (bMesh[boundaryPatchID]);
                // Get the neighbour patch
                neighbCells.append((patch.neighbPatch()).faceCells()[faces[faceI]]);
            }
        }
    }
    
    return neighbCells;
}
```

This function can get all the neighbor cells, however within the sort stencil function the distance to the reference cell is calculated with the absolute position of the cells, which does not know that some cells are collected over the cyclic boundary. This could in theory be also solved by tracking an additional list in which the positions of the cells in the stencil list are saved. For cells that are collected over a cyclic boundary the `transform()` function of `cyclicPolyPatch` could be used to get the correct, virtual, cell position. 
However, for the geometryWENO functions, especially the `transformIntegral` function the cells on the mesh are required. Here, all points that are calculated with the `polyMeshTetDecomposition::cellTetIndices`would needed to be transformed as well. Yet, at this point in the code it is not known from which cyclic boundary the transformation should be taken, unless this information is tracked as well. 

In conclusion. adding the function to calculate the stencil list and matrices over cyclic boundaries would require a lot of communication and tracking overhead, which introduce a lot of complexity, and possible errors. 
Therefore, I have decided to not include this function.

If anyone sees, a simpler and effective way to integrate the stencil collection for coupled boundaries please post an idea here or generate a pull request if you have a proposed solution.

**Comment by linysh1997** [comment]
Hello, Jan.
Is the current code able to treat cyclic boundary condition? I am running some testing cases of 2D vortex in isentropic flow with cyclic boundary conditions using WENOExt, but just found this post.

**Reply by JanGaertner** [reply]
Hello,

the WENO scheme can be used with cyclic boundary conditions, however, it will loose some of its accuracy over the cyclic boundary. The issue with processor boundaries is solved using a reconstructed regional mesh. This however does not work for cyclic boundaries, especially if the cyclic boundaries are on the same processor. 

So in conclusion, yes it is an issue, but it may not affect your results as much, best to give it a try. 

Just FYI, I just ran a vortex test case of cerfacs: https://elearning.cerfacs.fr/pdfs/numerical/TestCaseVortex2D.pdf with cyclic boundaries and only the WENO scheme can preserve some of the vortex shape, see attached plots and images. 

![compressible](https://github.com/WENO-OF/WENOEXT/assets/33832936/ef2c92c0-d0c0-4bb2-a21f-f6b41db4669b)

## Linear
![compressible_linear](https://github.com/WENO-OF/WENOEXT/assets/33832936/5482e04e-93b2-45dc-b89c-8d341e7d857f)

## Limited Linear
![compressible_limitedLinear](https://github.com/WENO-OF/WENOEXT/assets/33832936/1a940571-95ce-4a3e-b6e9-c7ea8fde1716)

## WENO 3 1 
![compressible_WENO](https://github.com/WENO-OF/WENOEXT/assets/33832936/a45aa53a-1b87-4dd6-b550-e0ff9794d8a6)

**Reply by linysh1997** [reply]
Thank you so much!
